eth/filters/api.go

```diff
@@ -379,6 +391,63 @@ func (s *PublicFilterAPI) NewFilter(args NewFilterArgs) (string, error) {
 	return externalId, nil
 }
 
+// logsSubscription implements the event.Event interface and wraps a log subscription.
+type logsSubscription struct {
+	api      *PublicFilterAPI
+	filterId string
+	logs     chan *event.Event
+}
+
+// Chan returns the channel where logs are published on for this subscription.
+func (sub *logsSubscription) Chan() <-chan *event.Event {
+	return sub.logs
+}
+
+// Unsubscribe will uninstall the filter and closes the emit channel
+func (sub *logsSubscription) Unsubscribe() {
+	if sub.api.UninstallFilter(sub.filterId) {
+		close(sub.logs)
+	}
+}
+
+// Logs creates a subscription and emits logs that match the given criteria.
+func (s *PublicFilterAPI) Logs(args NewFilterArgs) (rpc.Subscription, error) {
+	externalId, err := newFilterId()
+	if err != nil {
+		return rpc.Subscription{}, err
+	}
+
+	logs := make(chan *event.Event)
+
+	// don't use earliest and latest since these are not applicable for subscriptions
+	var id int
+	if len(args.Addresses) > 0 {
+		id, err = s.newLogFilter(-1, -1, args.Addresses, args.Topics, logs)
+	} else {
+		id, err = s.newLogFilter(-1, -1, nil, args.Topics, logs)
+	}
+
+	if err != nil {
+		return rpc.Subscription{}, err
+	}
+
+	s.filterMapMu.Lock()
+	s.filterMapping[externalId] = id
```

Comment:
Could you explain this? Subscription logs are never requested but pushed. This approach is the same as for the polling filters. If there is an issue it might be there. But the filter id that is necessary to request logs is only send to the client after the function returns and thus part of the filterMapping collection.